precompiles/utils/macro/src/as_precompile_name.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright 2019-2022 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::*;
+use syn::{GenericArgument, Type};
+
+pub fn main(_: TokenStream, input: TokenStream) -> TokenStream {
+	let item = parse_macro_input!(input as ItemType);
+
+	let ItemType {
+		attrs,
+		vis,
+		type_token,
+		ident,
+		generics,
+		eq_token,
+		ty,
+		semi_token,
+	} = item;
+
+	if let Type::Tuple(ref type_tuple) = *ty {
+		let variants: Vec<(Ident, u64)> = type_tuple
+			.elems
+			.iter()
+			.filter_map(as_precompile_name_variant)
+			.collect();
+
+		let ident_expressions: Vec<&Ident> = variants.iter().map(|(ident, _)| ident).collect();
+		let variant_expressions: Vec<&u64> = variants.iter().map(|(_, id)| id).collect();
+
+		(quote! {
+			#(#attrs)*
+			#vis #type_token #ident #generics #eq_token #ty #semi_token
+
+			#[derive(num_enum::TryFromPrimitive, num_enum::IntoPrimitive)]
+			#[repr(u64)]
+			pub enum PrecompileName {
+				#(
+					#ident_expressions = #variant_expressions,
+				)*
+			}
+
+			impl PrecompileName {
+				pub fn from_address(address: sp_core::H160) -> Option<Self> {
+					let _u64 = address.to_low_u64_be();
+					if address == sp_core::H160::from_low_u64_be(_u64) {
+						use num_enum::TryFromPrimitive;
+						Self::try_from_primitive(_u64).ok()
+					} else {
+						None
+					}
+				}
+			}
+		})
+		.into()
+	} else {
+		return quote_spanned! {
+			ty.span() => compile_error("Expected tuple");
+		}
+		.into();
+	}
+}
+
+fn as_precompile_name_variant(type_: &Type) -> Option<(Ident, u64)> {
+	if let Type::Path(type_path) = type_ {
+		if let Some(path_segment) = type_path.path.segments.last() {
```

Comment:
Maybe check that it is a `PrecompileAt`? Currently we only have that but that may be a source of surprises if we add a new one or refactor/rename `PrecompileAt` in the future.

Reply:
Done